akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -145,6 +145,16 @@ final class Flow[-In, +Out, +Mat](
   override def mapMaterializedValue[Mat2](f: Mat => Mat2): ReprMat[Out, Mat2] =
     new Flow(traversalBuilder.transformMat(f), shape)
 
+  /**
+   * Materializes this [[Flow]], immediately returning (1) its materialized value, and (2) a newly materialized [[Flow]].
+   * The returned flow is partial materialized and do not support multiple times materialization.
+   */
+  def preMaterialize()(implicit materializer: Materializer): (Mat, ReprMat[Out, NotUsed]) = {
+    val ((sub, mat), pub) =
+      Source.asSubscriber[In].viaMat(this)(Keep.both).toMat(Sink.asPublisher(false))(Keep.both).run()
+    (mat, Flow.fromSinkAndSource(Sink.fromSubscriber(sub), Source.fromPublisher(pub)))
```

Comment:
this can not be the `fromSinkAndSourceCoupled`

Reply:
I think a user wouold expect it to be coupled though, if not, a failure either upstream or downstream will not propagate to the other side once the stream is fully materialized.

Reply:
I see now, it propagates through the publisher/subscriber anyway and the extra propagation in fromSinkAndSource is not needed. 👍 .